### ISSUE TITLE
Trouble setting up irony mode with company and json compilation database

### ISSUE THREAD

**Issue by LefterisJP** [Sarcasm/irony-mode#137]
Hello,

First of all I would like to thank you for making irony-mode. It seems like a really nice package that greatly enhances the C/C++ editing experience.

I am an rtags user, but wanted to combine it with irony mode so as to get nice completions. Thus I tried to use company mode and company-irony. I obtained the packages from el-get. My setup is as follows:

>init-irony-mode.el:

``` elisp
(add-hook 'c++-mode-hook 'irony-mode)
(add-hook 'c-mode-hook 'irony-mode)
(add-hook 'objc-mode-hook 'irony-mode)

;; replace the `completion-at-point' and `complete-symbol' bindings in
;; irony-mode's buffers by irony-mode's function
(defun my-irony-mode-hook ()
  (define-key irony-mode-map [remap completion-at-point]
    'irony-completion-at-point-async)
  (define-key irony-mode-map [remap complete-symbol]
    'irony-completion-at-point-async))
(add-hook 'irony-mode-hook 'my-irony-mode-hook)
(add-hook 'irony-mode-hook 'irony-cdb-autosetup-compile-options)
```

>init-company-irony.el

``` elisp
(eval-after-load 'company
  '(add-to-list 'company-backends 'company-irony))

;; (optional) adds CC special commands to `company-begin-commands' in order to
;; trigger completion at interesting places, such as after scope operator
;;     std::|
(add-hook 'irony-mode-hook 'company-irony-setup-begin-commands)
```
Nothing fancy. Just what the wikis say. I also got `(add-hook 'after-init-hook 'global-company-mode)` somewhere so that I activate company mode.

I am using json compilation databases for all of my projects. I use [a package](https://github.com/LefterisJP/malinka) that I have created to act as a glue between other emacs packages and create the json compilation database. I know that the database is valid since rtags server gets it as input and correctly navigates the projects.

So I have this json compilation named compile_commands.json in the root of the project. I suspect that irony mode actually reads it since I see something that looks like the whole project's compile flags, include directories and files when I execute `irony-cdb-menu`.

But the problem is that whenever I try to use any autocomplete or anything at all, all I get is an error message:
>error in process filter: irony-completion-at-point: Wrong type argument: sequencep, libclang:

I also tried to describe the current company backend variable and it seems to be nil. This is funny because `company-backends` list has company-irony as the first one in the list. I tried to insert `company-irony` as the only element of company-backends but I got the same result.

While debugging I tried to remove `(add-hook 'irony-mode-hook 'irony-cdb-autosetup-compile-options)` from the setup. Then completion popup actually appeared but it was totally random. Seemed to have no proper source. Could the issue be with the way that irony autodetects the compilation database?

Do you have any idea what may be happening? IF you need any more information about my setup I will gladly provide.

**Comment by LefterisJP** [comment]
After some changes to my setup I can't reproduce it properly anymore. I still have issues with irony not picking the compilation database sometimes but it may be that it's a bit slower than I would have wanted because it does pick completions up after some time.

When I find the time I will need to debug it more and see if I still get any libclang crushes under any circumstances. If not then this issue may no longer be an issue.

**Comment by Sarcasm** [comment]
> After some changes to my setup I can't reproduce it properly anymore. I still have issues with irony not picking the compilation database sometimes but it may be that it's a bit slower than I would have wanted because it does pick completions up after some time.

irony-mode may take some time to load the completion database when you open the buffer but this is not something done in background so you would notice it if it's too long (and you wouldn't be able to perform code-completion or things like that until it's done anyway). So in theory, either the compile options are right or they aren't but it is not something that should change after some time. The issue may be somewhere else, let me know of your findings if it happens (potentially in another issue if this one is closed).

**Comment by DanSchoppe** [comment]
Sarcasm,

I'm running into a similar issue as LefterisJP was describing.  It sounds like LefterisJP was unable to recreate the issue any longer so maybe I can help track down the issue.  If my problem turns out to be unrelated I can just create a new thread.

Installation: I installed company, company-irony, and irony-mode packages from Melpa.  I updated my init.el to include the suggested entries for the three packages, exactly how LefterisJP listed in his first post.  I built and installed irony server pointing to my homebrew installation of llvm 3.5.1 by adding the -DCMAKE_PREFIX_PATH=[path/to/llvm/3.5.1] to the cmake args.

Project: Makefile project, so I used bear to generate a JSON database containing compile flags.  The compile_commands.json file looks like it has a list of flags for each object as I would expect.  Just like LefterisJP, I am successfully using this JSON database with RTags.

When I run irony-completion-at-point-async in a project file, I get back "Wrong type argument: sequencep, libclang:"  After toggling on debug on error I get a backtrace with the same contents as LefterisJP listed earlier.

Here's where I start to get suspicious... When I inspect irony--compile-options (after running irony-completion-at-point-async), only the first entry from the compile_commands.json file is listed in the variable.  Perhaps LefterisJP's compile_commands.json file changed due to another generation tool or from changes to his project that fixed the irony--compile-options issue.

I also tested the irony-server in a bash terminal (not sure if this is relevant or helpful):
<code>./irony-server -d complete [path/to/source/file.cpp] 100 7 RET RET

execute: Command{action=Command::Complete, file='[path/to/source/file.cpp]', line=100, column=7, flags=[], unsavedFiles.count=0, opt=off}
Assertion failed: (!T->isBeingDefined() && "Cannot serialize in the middle of a type definition"), function VisitTagType, file ASTWriter.cpp, line 277.
libclang: crash detected during reparsing
Assertion failed: (OutputFiles.empty() && "Still output files in flight?"), function ~CompilerInstance, file CompilerInstance.cpp, line 62.
Abort trap: 6</code>

Any thoughts or help is appreciated.  Thanks.

**Comment by Sarcasm** [comment]
@DanSchoppe: for now I don't think it's the same issue, let us continue here: #191

**Comment by egarrulo** [comment]
Thanks for sharing this package.  My understanding is that `irony-cdb-autosetup-compile-options` is not available in Irony 0.1.2, the stable version of Irony.  I am writing this to let other newcomers know.

**Comment by Sarcasm** [comment]
I'm closing this issue.

@egarrulo you are right, the compilation database stuff isn't ready yet to have a proper release, I need to rework it.
